### ISSUE TITLE
Non-saturating arithmetic on saturating types

### ISSUE THREAD

**Issue by kornelski** [rust-lang/rust-clippy#15328]
### What it does

I've had a tuple with a `Saturating<u8>` field, and I've written `num.0 += 1` thinking I'm accessing a field of my tuple, but actually it was the number itself, and I accidentally accessed the inner value of the `Saturating` wrapper, and ended up having overflowing arithmetic instead.

### Advantage

Ensures the wrapper type isn't accidentally bypassed. The whole point of the wrapper is to ensure it's used.

### Drawbacks

Someone might want to perform direct arithmetic on the inner type when knowing it won't overflow, but in that case maybe an explicit `allow(lint)` wouldn't be too bad to communicate the intention. Or they could just use the wrapper type when the arithmetic is obviously in bounds and the optimizer can remove range checks.

### Example

```rust
    let mut n = Saturating(0u8);
    n.0 += 100;
```

Could be written as:

```rust
    let mut n = Saturating(0u8);
    n += 100;
```

**Comment by kornelski** [comment]
Most likely same applies to `Wrapping`, although that is less disastrous in release mode.
